TESTS/mbed_hal_fpga_ci_test_shield/analogout/main.cpp

```diff
@@ -0,0 +1,126 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if !DEVICE_ANALOGOUT
+#error [NOT_SUPPORTED] Analog out not supported for this target
+#elif !COMPONENT_FPGA_CI_TEST_SHIELD
+#error [NOT_SUPPORTED] FPGA CI Test Shield is needed to run this test
+#else
+
+#include "utest/utest.h"
+#include "unity/unity.h"
+#include "greentea-client/test_env.h"
+#include <inttypes.h>
+#include "mbed.h"
+
+using namespace utest::v1;
+
+#include "MbedTester.h"
+#include "pinmap.h"
+#include "test_utils.h"
+
+#if !DEVICE_ANALOGOUT
+#error [NOT_SUPPORTED] Analog out not supported for this target
+#endif
+
+#define analogout_debug_printf(...)
+
+#define DELTA_FLOAT                     0.01f    // 1%
+
+const PinList *form_factor = pinmap_ff_default_pins();
+const PinList *restricted = pinmap_restricted_pins();
+
+MbedTester tester(form_factor, restricted);
+
+void analogout_init_free(PinName pin)
+{
+    dac_t analogout;
+
+    analogout_debug_printf("Analog output init/free test on pin=%s (%i)\r\n", pinmap_ff_default_pin_to_string(pin), pin);
+
+    analogout_init(&analogout, pin);
+    analogout_free(&analogout);
+}
+
+void analogout_test(PinName pin)
+{
+    analogout_debug_printf("Analog input test on pin %s (%i)\r\n", pinmap_ff_default_pin_to_string(pin), pin);
+
+    tester.reset();
+    tester.peripherals_reset();
+
+    /* Test analog input */
+
+    dac_t analogout;
+    analogout_init(&analogout, pin);
+
+    tester.set_sample_adc(true);//begin ADC sampling on the FPGA
+
+    float anin;
+    float i;
+    for (i = 0.0f; i < 0.304f; i += 0.0303f) {//0V-1V, float
+        analogout_write(&analogout, i);
+        anin = tester.get_analog_in();
+        TEST_ASSERT_FLOAT_WITHIN(DELTA_FLOAT, (i * 3.3f), anin);
+    }
+
+    i = 0.0f;
+    for (uint16_t i_d16 = 0; i_d16 < 19851; i_d16 += 1985) {//0V-1V, 16-bit
+        analogout_write_u16(&analogout, i_d16);
+        anin = tester.get_analog_in();
+        TEST_ASSERT_FLOAT_WITHIN(DELTA_FLOAT, (i * 3.3f), anin);
+        i += 0.0303f;
+    }
+
+    analogout_free(&analogout);
+
+    tester.set_sample_adc(false);//stop ADC sampling on the FPGA
+
+    // power analysis
+    // uint64_t sum;
+    // uint32_t samples;
+    // uint64_t cycles;
+    // tester.get_anin_sum_samples_cycles(0, &sum, &samples, &cycles);
+    // printf("ANIN0\r\n");
+    // printf("Sum: %llu\r\n", sum);
+    // printf("Num power samples: %d\r\n", samples);
+    // printf("Num power cycles: %llu\r\n", cycles);
+    // printf("ANIN0 voltage: %.6fV\r\n", tester.get_anin_voltage(0));
```

Comment:
Can we remove these lines not used?

Reply:
I think that this code has been added for debugging. Maybe I can wrap this code around `#if DEBUG` directive?

Reply:
Fixed